core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1064 @@ def _convert_states_v50_dict_to_v51_dict(
+                recc_exp_ids = state_dict['interaction'][
+                    'customization_args']['recommendedExplorationIds']['value']
+                # All recommended explorations should be valid.
+                valid_recc_exp_ids = []
+                for recc_exp in recc_exp_ids:
+                    if recc_exp in exp_ids:
+                        valid_recc_exp_ids.append(recc_exp)
+                recc_exp_ids = valid_recc_exp_ids
+                state_dict['interaction'][
+                'customization_args']['recommendedExplorationIds'][
+                    'value'] = valid_recc_exp_ids
+
+                # Should be at most 3 recommended explorations.
+                if len(recc_exp_ids) > 3:
+                    recc_exp_ids = recc_exp_ids[0:3]
+                    state_dict['interaction']['customization_args'][
+                        'recommendedExplorationIds']['value'] = recc_exp_ids
+
+            # NumericInput Interaction.
+            if state_dict['interaction']['id'] == 'NumericInput':
+                answer_groups = state_dict['interaction']['answer_groups']
+                # Each answer group should not be a subset of any answer
+                # group that comes before it.
+                answer_groups = cls._numeric_ans_group_should_not_be_subset(
+                    answer_groups)
+                for answer_group in answer_groups:
+                    for rule_spec in answer_group['rule_specs']:
+                        # For x in [a-b, a+b], b must be a positive value.
+                        if rule_spec['rule_type'] == 'IsWithinTolerance':
+                            tol = rule_spec['inputs']['tol']
+                            if tol <= 0:
+                                rule_spec['inputs']['tol'] = abs(tol)
+
+                        # For x in [a, b], a must not be greater than b.
+                        if rule_spec['rule_type'] == 'IsInclusivelyBetween':
+                            if (
+                                rule_spec['inputs']['a'] > rule_spec[
+                                    'inputs']['b']
+                            ):
+                                rule_spec['inputs']['a'], rule_spec[
+                                    'inputs']['b'] = rule_spec['inputs'][
+                                        'b'], rule_spec['inputs']['a']
+
+                state_dict['interaction']['answer_groups'] = answer_groups
+
+            # FractionInput Interaction.
+            if state_dict['interaction']['id'] == 'FractionInput':
+                # All rules should have solutions that do not match
+                # previous rules' solutions.
+                answer_groups = state_dict['interaction']['answer_groups']
+                answer_groups = cls._fraction_rules_should_not_match_prev_rule(
+                    answer_groups
+                )
+                state_dict['interaction']['answer_groups'] = answer_groups
+
+            # MultipleChoiceInput Interaction.
```

Comment:
MultipleChoice part below is also half decomposed, right?

Reply:
Updated